score_ui.gd

```diff
@@ -0,0 +1,22 @@
+extends HBoxContainer
+class_name ScoreUI
+
+
+@export var game: Game = null
+
+@onready var other_score: Label = $OtherScore
+@onready var player_score: Label = $PlayerScore
+
+
+func on_game_other_score_changed(value: int) -> void:
+	self.other_score.set_text(str(value))
+
+
+func on_game_player_score_changed(value: int) -> void:
+	self.player_score.set_text(str(value))
+
+
+func _ready() -> void:
+	assert(game != null, "A Game node must be provided in the inspector!")
+	game.other_score_changed.connect(on_game_other_score_changed)
+	game.player_score_changed.connect(on_game_player_score_changed)
```

Comment:
I'm not too keen on this design, I prefer signal connections to be from children to parents and method calls to be from parent to children. Since the parent already has a reference to ScoreUI, it could call something like score_ui.increment(player:bool) which would increment the proper label without having the signals. I should probably move the scores into the score UI entirely, removing them from game.gd.

Reply:
Game doesn't have a ref to the UI in this PR and here is why:

Game doesn't care about ScoreUI. If ScoreUI is not present on Game, Game will still work and the score will still be updated in it.
When the score is updated it sends signals to notify the player and enemy (aka "other") from these changes.
It doesn't care about whoever listens to it and if 0 or N things are listening to it there would be no problem at all for Game.

ScoreUI is dumb but it needs data to **display** something. It **doesn't manage / calculate** anything so it won't increment anything.
It only relies on Game to provide the data by listening to the score change signals (EG: `player_score_changed`) and displaying it.

Now if Game had a ref to ScoreUI but this scene is not present in Game, it'd very likely crash since the ref would break.

If you want to decouple the score from Game the same way I did, I suggest you create ScoreManager scene which holds the signals you need (the same than the score ones contained in Game) and make ScoreUI listen to ScoreManager.
All you'd have to do is to change the game variable + its type, then click in the editor in the exported field of ScoreUI and choose ScoreManager.
You'd then have to make ScoreManager listen to signals when goals are scored which implies signals in Game should be created like: `signal player_scored`


That's the reasoning behind  this approach.


Here is an interesting article about all of that:
https://kidscancode.org/godot_recipes/4.x/basics/node_communication/index.html#signal-example

Reply:
Indeed, I've read this article and that was my point: 

![image](https://github.com/Xyloph/20GC_Game1/assets/611890/549074f0-ade6-4766-a4ee-b52e9187b23d)

In this particular scene, Game is the parent and ScoreUI is a child.

Reply:
However I also reworked the score in https://github.com/Xyloph/20GC_Game1/commit/af0f737fc66c7ef45ab5b7966b1c207fef02567b where ScoreUI was the sole responsible for keeping the score.

Reply:
Indeed yes I read it again and it seems like even though I apply these principles (not so much in this particular case) I also added my own twist which is not mentionned in this article,  sorry about that.

The idea was to decouple things a bit more and make sure things don't break with missing references of the child in the parent.
By making components which need a parent to work.
But I'd set the parent ref via the editor or use get_parent() to attach the component to another node.

My logic is if there is no parent given in the ref (then it'd most likely crash) then that means this component should most likely not exist.

That's why I took a little detour from the principles listed in this article.

Even though you did it differently from my approach with ScoreUI  (which is to **display** data from an exterior source and **not managing / calculating** anything) I hope it gave you at least some insights! :)

Reply:
yes, and thanks for the insight! I understand that having a ScoreManager like you mentioned would make the principle of call down/signal up more applicable. Ideally I shouldn't have one big Game master scene with level elements in them but rather one Main scene that has reusable game nodes and UI nodes that work separately from one another and would be peers rather than parent/child